flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -332,11 +346,18 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
                 context.getCheckpointId(),
                 currentTransactionHolder);
 
-        preCommit(currentTransactionHolder.handle);
-        pendingCommitTransactions.put(checkpointId, currentTransactionHolder);
-        LOG.debug("{} - stored pending transactions {}", name(), pendingCommitTransactions);
+        if (!currentTransactionHolder.equals(TransactionHolder.empty())) {
```

Comment:
As I understand, we don't have the test which covers this condition. I mean it makes sense to write the test when currentTransactionHolder is empty. What do you think?

Reply:
That is a good point, the problem with this is from the test-result point of view, the results are the same. 

But reading the test code again, I think it could be tested by adding an additional checkpoint after the finish.

In that case, if the previous empty transaction holder is pre-committed, an exception will be thrown because its TXN is null.

Additionally, I can add another test to make sure no data is expected to injected after finish (the checkNotNull does work.

WDYT?